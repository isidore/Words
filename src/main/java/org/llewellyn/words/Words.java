package org.llewellyn.words;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

import com.spun.util.ObjectUtils;
import com.spun.util.StringUtils;
import org.lambda.query.Queryable;

public class Words
{

    public static Queryable<String> get() throws Error {
        return importWords("words.txt");
    }

    public static Queryable<String> importWords(String filename)
    {
        Queryable<String> words = new Queryable<>(String.class);
        InputStream resourceAsStream = Words.class.getResourceAsStream("../../../"+filename);
        BufferedReader reader = new BufferedReader(new InputStreamReader(resourceAsStream));
        try
        {
            while (reader.ready())
            {
                String word = reader.readLine().trim();
                if (!StringUtils.isEmpty(word))
                {
                    words.add(word);
                }
            }
            reader.close();
        }
        catch (IOException e)
        {
            ObjectUtils.throwAsError(e);
        }
        return words;
    }
}