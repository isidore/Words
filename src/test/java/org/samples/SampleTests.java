package org.samples;


import org.approvaltests.Approvals;
import org.approvaltests.reporters.UseReporter;
import org.approvaltests.reporters.macosx.DiffMergeReporter;
import org.junit.jupiter.api.Test;
import org.lambda.functions.Function1;
import org.lambda.query.OrderBy;
import org.lambda.query.Queryable;
import org.llewellyn.words.Words;

import java.util.Locale;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;

@UseReporter(DiffMergeReporter.class)
public class SampleTests
{

  //@Test
  public void test()
  {
      Queryable<String> words = wordle("easclin")
             .where(letterAt('o', 2))
              //.where(letterAt('i', 3))
              .where(letterNotAt('r',4))
              .where(letterNotAt('t', 1))
              ;
      Approvals.verifyAll("", words);
  }

    @Test
    public void testLetterDistribution()
    {
        Queryable<Map.Entry<String, Queryable<String>>> words = getLetterDistribution();
        Approvals.verifyAll("", words, m -> String.format("%s = %s", m.getKey(), m.getValue().size()));
    }

    private Queryable<Map.Entry<String, Queryable<String>>> getLetterDistribution() {
        Queryable<Map.Entry<String, Queryable<String>>> words = wordle(" ")
                .selectManyArray(w -> w.split("")).groupBy(l -> l)
                .orderBy(OrderBy.Order.Descending, m -> m.getValue().size())
                ;
        return words;
    }

    @Test
    public void testSecondBest()
    {
        var distribution = getLetterDistribution();


        var words = wordle("tears")
               // .where(w -> include.all(l -> w.contains(l)))
                .where(containsNoDuplicates())
                .orderBy(OrderBy.Order.Descending, w -> toLetters(w).sum(l -> distribution.where(d -> d.getKey().equals(l)).first().getValue().size()))
                ;
        words = take(words, 20);
        Approvals.verifyAll("", words);
    }

    private Queryable<String> take(Queryable<String> words, int number) {
        return  Queryable.as(words.subList(0, number),words.getType());
    }

    private Queryable<String> toLetters(String possible) {
        return Queryable.as(possible.split(""));
    }

    private Function1<String, Boolean> containsNoDuplicates() {
      return w -> toLetters(w).groupBy(l -> l).size() ==5 ;
    }
    private Function1<String, Boolean> letterAt(char letter, int position) {
      return w -> w.charAt(position-1) == letter;
    }

    private Function1<String, Boolean> letterNotAt(char letter, int position) {
        return w -> w.charAt(position-1) != letter && w.contains("" + letter);
    }
    private Queryable<String> wordle(String excludingLetters) {
        var exclude = toLetters(excludingLetters);
        return Words.get()
                .where(w -> w.length() == 5)
                .where(w -> w.equals(w.toLowerCase(Locale.ROOT)))
                .where(w -> exclude.all(l -> !w.contains(l)));
    }
}
