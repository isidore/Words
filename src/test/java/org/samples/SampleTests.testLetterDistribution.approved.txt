a = 4467
e = 4255
r = 3043
o = 2801
i = 2581
s = 2383
t = 2381
l = 2368
n = 2214
u = 1881
y = 1605
c = 1546
d = 1399
h = 1323
m = 1301
p = 1293
b = 1162
g = 1102
k = 882
w = 685
f = 661
v = 466
z = 250
x = 189
j = 163
q = 84
